README.md
# eastwood - a Clojure lint tool

[![Dependencies Status](http://jarkeeper.com/jonase/eastwood/status.svg)](http://jarkeeper.com/jonase/eastwood)

<img src="doc/Clint_Eastwood_-_1960s_small.jpg"
 alt="Picture of Clint Eastwood in 'A Fistful of Dollars' (1964)" title="Clint Eastwood in 'A Fistful of Dollars' (1964)"
 align="right" />

> "Now remember, things look bad and it looks like you're not gonna
> make it, then you gotta get mean.  I mean plumb, mad-dog mean.
> 'Cause if you lose your head and you give up then you neither live
> nor win.  That's just the way it is."
> - Josey Wales, played by Clint Eastwood in "The Outlaw Josey Wales"

Eastwood is a Clojure
[lint](http://en.wikipedia.org/wiki/Lint_%28software%29) tool that
uses the [tools.analyzer](https://github.com/clojure/tools.analyzer)
and
[tools.analyzer.jvm](https://github.com/clojure/tools.analyzer.jvm)
libraries to inspect namespaces and report possible problems.

Clojure version compatibility:

* Eastwood supports only Clojure on Java, not ClojureScript or
  Clojure/CLR.

* Clojure 1.8.0 - Use Eastwood 0.2.2 or later.  There are known
  problems using Eastwood 0.2.1 and earlier with Clojure 1.8.0.

* Clojure 1.5.1, 1.6.0, or 1.7.0 - Many versions of Eastwood have been
  tested with these, up through Eastwood 0.2.3.

* Clojure 1.4.0 - Use Eastwood 0.1.5 or earlier.

The `.cljc` files introduced in Clojure 1.7.0 are linted, starting
with Eastwood version 0.2.4.  Earlier Eastwood versions ignore such
files.


## Installation & Quick usage

Eastwood can be run from within a REPL, regardless of which build
tools you may use.  See the [instructions
here](#running-eastwood-in-a-repl).

Eastwood can be run from the command line as a
[Leiningen](http://leiningen.org) plugin.  As a Leiningen plugin,
Eastwood has been tested most with Leiningen versions 2.4 and later.
There is a known bug in Leiningen 2.6.0 where many plugins, including
Eastwood, often cause exceptions to be thrown -- Leiningen 2.6.1 was
released to fix that problem.

Merge the following into your `$HOME/.lein/profiles.clj` file:

```clojure
{:user {:plugins [[kirasystems/eastwood "0.2.5"]] }}
```

To run Eastwood with the default set of lint warnings on all of the
Clojure files in the source _and_ test paths of your project, use the
command:

    $ lein eastwood

WARNING: If loading your code (particularly test files) causes side
effects like writing files, opening connections to servers, modifying
databases, etc., running Eastwood on your code will do that, too.
Eastwood is _no less dangerous_ than loading your code, and should be
no more dangerous.  To confine linting to files in your
`:source-paths`, use this command instead:

    $ lein eastwood "{:namespaces [:source-paths]}"

If it is not obvious what a warning message means, _please_ check the
next section, which has a `[more]` link for each type of warning.
Most types of warning messages have a page or more of text describing
the warning, why it occurs, and sometimes suggestions on what you can
do about it.  Also note that there are several types of warnings
marked as '(disabled)', meaning that by default no such warnings will
be checked for.  You may wish to enable those for your project.  See
the [Usage](#usage) section for options to enable or disable types of
warnings for your entire project.

See [Editor Support](#editor-support) below for instructions on using
a text editor or IDE to quickly take you to the file, line, and column
of each warning message.

See the [Usage](#usage) section below for more notes on side effects
in test code, and instructions on [running Eastwood in a REPL
session](#running-eastwood-in-a-repl).

Eastwood can only finish linting a file if Clojure itself can compile
it (unlike some other lint tools, which try to give meaningful error
messages for programs with syntax errors).  It is recommended to use a
command like `lein check` to check for compiler errors before running
Eastwood.  Even better, `lein test` will compile files in your source
paths and test paths, not merely your source paths as `lein check`
does.

If you run Eastwood from a `lein` command line, it is perfectly normal
to see the message `Subprocess failed` at the end if either the
warning or exception thrown counts are not 0.  Eastwood exits with a
non-0 [exit status](http://en.wikipedia.org/wiki/Exit_status) in this
situation, so that shell scripts or build tools running Eastwood will
have a simple way to check that something was not perfect.  If
Eastwood quits due to some internal error that throws an exception,
you will typically see much more voluminous output about what went
wrong, often including a stack trace.

See section [For Eastwood developers](#for-eastwood-developers) below
for instructions on trying out the latest unreleased version of
Eastwood.


## What's there?

Eastwood warns when it finds the following kinds of things.  Each
keyword below is the name of the "linter".  That name can be used on
the command line to enable or disable the linter.  All linters are
enabled by default unless they have '(disabled)' after their name.

| Linter name | Description | Docs |
| ----------- | ----------- | ---- |
| no name* | Inconsistencies between file names and the namespaces declared within them.  * Cannot be disabled. (added 0.1.1) | [[more]](#check-consistency-of-namespace-and-file-names) |
| `:bad-arglists` | Function/macro `:arglists` metadata that does not match the number of args it is defined with (added 0.1.1). | [[more]](#bad-arglists) |
| `:constant-test` | A test expression always evaluates as true, or always false (added 0.2.0). | [[more]](#constant-test) |
| `:def-in-def` | def's nested inside other def's. | [[more]](#def-in-def) |
| `:deprecations` | Deprecated Clojure Vars, and deprecated Java constructors, methods, and fields. | [[more]](#deprecations) |
| `:keyword-typos` (disabled) | Keyword names that may be typos because they occur only once in the source code and are slight variations on other keywords. | [[more]](#keyword-typos) |
| `:local-shadows-var` | A local name, e.g. a function arg or let binding, has the same name as a global Var, and is called as a function (added 0.1.5). | [[more]](#local-shadows-var) |
| `:misplaced-docstrings` | Function or macro doc strings placed after the argument vector, instead of before the argument vector where they belong. | [[more]](#misplaced-docstrings) |
| `:no-ns-form-found` | Warn about Clojure files where no `ns` form could be found (added 0.2.0). | [[more]](#no-ns-form-found) |
| `:non-clojure-file` (disabled) | Warn about files that will not be linted because they are not Clojure source files, i.e. their name does not end with '.clj' (added 0.2.0). | [[more]](#non-clojure-file) |
| `:redefd-vars` | Redefinitions of the same name in the same namespace. | [[more]](#redefd-vars) |
| `:suspicious-expression` | Suspicious expressions that appear incorrect, because they always return trivial values. | [[more]](#suspicious-expression) |
| `:suspicious-test` | Tests using `clojure.test` that may be written incorrectly. | [[more]](#suspicious-test) |
| `:unlimited-use` | Unlimited `(:use ...)` without `:refer` or `:only` to limit the symbols referred by it. | [[more]](#unlimited-use) |
| `:unused-fn-args` (disabled) | Unused function arguments. | [[more]](#unused-fn-args) |
| `:unused-locals` (disabled) | Symbols bound with `let` or `loop` that are never used (added 0.2.0). | [[more]](#unused-locals) |
| `:unused-meta-on-macro` | Metadata on a macro invocation is ignored by Clojure (added 0.2.0). | [[more]](#unused-meta-on-macro) |
| `:unused-namespaces` (disabled) | Warn if a namespace is given in an `ns` form after `:use` or `:require`, but no Vars within the namespace are ever mentioned. | [[more]](#unused-namespaces) |
| `:unused-private-vars` (disabled) | Unused private vars (updated in version 0.2.0). | [[more]](#unused-private-vars) |
| `:unused-ret-vals` and `:unused-ret-vals-in-try` | Unused values, including unused return values of pure functions, and some others functions where it rarely makes sense to discard its return value. | [[more]](#unused-ret-vals) |
| `:wrong-arity` | Function calls that seem to have the wrong number of arguments. | [[more]](#wrong-arity) |
| `:wrong-ns-form` | ns forms containing incorrect syntax or options (added 0.2.1). | [[more]](#wrong-ns-form) |
| `:wrong-pre-post` | function has preconditions or postconditions that are likely incorrect (added 0.2.2). | [[more]](#wrong-pre-post) |
| `:wrong-tag` | An incorrect type tag for which the Clojure compiler does not give an error (added 0.1.5). | [[more]](#wrong-tag) |

The following table gives some additional detail about each linter.

The 'debug' column indicates whether extra debug messages about a
linter's warnings can be enabled via the `:debug-warning` option.
This option can be given a value of `true` to enable all such
warnings, or it can be a set of keywords that also enables additional
details to be printed.  The only keyword currently supported in this
set is `:ast`, which prints AST contents related to issued warnings
for most linters that implement `:debug-warning`.

The 'suppress' column indicates whether warnings produced by the
linter can be selectively disabled via Eastwood config files.  See
[Eastwood config files](#eastwood-config-files) for more details.

| Linter name | debug | suppress |
| ----------- | ----- | -------- |
| no name*                 |  |  |
| `:bad-arglists`          |  |  |
| `:constant-test`         | yes | yes |
| `:def-in-def`            |  |  |
| `:deprecations`          |  |  |
| `:keyword-typos`         |  |  |
| `:local-shadows-var`     |  |  |
| `:misplaced-docstrings`  |  |  |
| `:no-ns-form-found`      |  |  |
| `:non-clojure-file`      |  |  |
| `:redefd-vars`           | yes | yes |
| `:suspicious-expression` | yes, for those involving macros | yes |
| `:suspicious-test`       |  |  |
| `:unlimited-use`         |  |  |
| `:unused-fn-args`        |  |  |
| `:unused-locals`         |  |  |
| `:unused-meta-on-macro`  |  |  |
| `:unused-namespaces`     |  |  |
| `:unused-private-vars`   |  |  |
| `:unused-ret-vals` and `:unused-ret-vals-in-try` | yes | yes |
| `:wrong-arity`           | yes | yes |
| `:wrong-ns-form`         |  |  |
| `:wrong-pre-post`        |  |  |
| `:wrong-tag`             |  |  |


## Usage

### From the command line as a Leiningen plugin

Running

    $ lein eastwood

in the root of your project will lint your project's namespaces -- all
of those in your `:source-paths` and `:test-paths` directories and
their subdirectories.  You can also lint individual namespaces in your
project, or your project's dependencies:

    $ lein eastwood "{:namespaces [compojure.handler compojure.core-test] :exclude-linters [:unlimited-use]}"
    == Eastwood 0.2.0 Clojure 1.5.1 JVM 1.7.0_45
    == Linting compojure.handler ==
    Entering directory `/Users/andy/clj/compojure'
    src/compojure/handler.clj:48:8: deprecations: Var '#'compojure.handler/api' is deprecated.
    == Linting compojure.core-test ==
    test/compojure/core_test.clj:112:21: suspicious-test: 'is' form has first arg that is a constant whose value is logical true.  This will always pass.  There is probably a mistake in this test
    test/compojure/core_test.clj:117:21: suspicious-test: 'is' form has first arg that is a constant whose value is logical true.  This will always pass.  There is probably a mistake in this test
    test/compojure/core_test.clj:109:1: constant-test: Test expression is always logical true or always logical false: false
    test/compojure/core_test.clj:109:1: constant-test: Test expression is always logical true or always logical false: true
    test/compojure/core_test.clj:114:1: constant-test: Test expression is always logical true or always logical false: false
    test/compojure/core_test.clj:114:1: constant-test: Test expression is always logical true or always logical false: true
    == Warnings: 7 (not including reflection warnings)  Exceptions thrown: 0
    Subprocess failed

Adding `:out "warn.txt"` to the options map will cause all of the
Eastwood warning lines and 'Entering directory' lines, but no others,
to be written to the file `warn.txt`.  This file is useful for
stepping through warnings as described in the [Editor
Support](#editor-support) section.

    # This works on bash shell in Linux and Mac OS X, and also in
    # Windows cmd shell
    $ lein eastwood "{:out \"warn.txt\"}"

    # This saves a little typing in bash shell, but does not work in
    # Windows cmd shell.  For all example command lines, you can use
    # single quotes in bash if you prefer.
    $ lein eastwood '{:out "warn.txt"}'

Available options for specifying namespaces and paths are:

* `:namespaces` Vector of namespaces to lint.  A keyword
  `:source-paths` in this vector will be replaced with a list of
  namespaces in your Leiningen `:source-paths` and their
  subdirectories.  These namespaces will be in an order that honors
  inter-namespace dependencies as determined by `:require` and `:use`
  keys in `ns` forms.  Similarly for a keyword `:test-paths`.  If you
  do not specify `:namespaces`, it defaults to `[:source-paths
  :test-paths]`.
* `:exclude-namespaces` Vector of namespaces to exclude.
  `:source-paths` and `:test-paths` may be used here as they can be
  for `:namespaces`.  Defaults to an empty list if you do not specify
  `:exclude-namespaces`.
* `:source-paths` is normally taken from your Leiningen `project.clj`
  file, which is `[ "src" ]` by default if not specified there.  You
  can also specify `:source-paths` in the Eastwood option map to
  override what Leiningen uses.
* `:test-paths` is similar in behavior to `:source-paths`, except it
  defaults to `[ "test" ]` if not specified in your `project.clj`
  file.

Linter names are given in the previous section.  Available options for
specifying which linters are enabled or disabled are:

* `:linters` Linters to use.  If not specified, same as `[:default]`,
  which is all linters except those documented as 'disabled by
  default'.
* `:exclude-linters` Linters to exclude
* `:add-linters` Linters to add.  The final list of linters is the set
  specified by `:linters`, taking away all in `:excluded-linters`,
  then adding all in `:add-linters`.

The keyword `:all` in any of the collections of linters listed above
will be replace with the collection of all linters.  The keyword
`:default` will be replaced with the collection of default linters.
Thus `:linters [:all]` enables all linters, even those disabled by
default, and `:linters [:all] :exclude-linters [:default]` enables
only those that are disabled by default.

Note that you can add Eastwood options to a user-wide Leiningen
`profiles.clj` file or to your project's `project.clj` file if you
wish.  See [How the Eastwood options map is
determined](#how-the-eastwood-options-map-is-determined) for more
details.

As mentioned in the [Installation & Quick
usage](#installation--quick-usage) section above, Eastwood causes any
and all side effects that loading the file would cause (e.g. by doing
`use` or `require` on the file's namespace).  Eastwood is able to find
potential problems in test code, too.  If you wish to use Eastwood on
test files without such side effects, consider modifying your tests so
that merely performing `require`/`use` on the files does not cause the
side effects.  If you can arrange things so that running your tests
requires loading the files and then calling some function(s) (e.g. as
tests written using
[`clojure.test/deftest`](http://clojure.github.io/clojure/#clojure.test)
do), then you can run Eastwood on those files without the side
effects.

If you know how to write tests using other Clojure test libraries
besides `clojure.test` so that merely loading the file does not run
the tests, please create an Issue for Eastwood on GitHub.

If you have a code base you do not trust to load, consider a sandbox,
throwaway virtual machine, etc.

New in Eastwood 0.1.5 is an option to force continuing to lint more
files even if an exception was thrown while analyzing or evaluating an
earlier file.  To do this, set the key `:continue-on-exception` to
`true` in the option map.  Without this option, the default behavior
is to stop linting if an exception is thrown.

Warning: Setting `:continue-on-exception` to true can cause exceptions
while analyzing later namespaces that would not otherwise occur.  For
example, if namespace `A` is analyzed first, and throws an exception
before function `A/foo` is defined, analyzing a later namespace `B`
that uses `A/foo` will throw an exception because it is undefined.

There are also options that enable printing of additional debug
messages during linting.  These are only intended for tracking down
the cause of errors in Eastwood.  You specify the key `:debug` with a
value that is a list or vector of keywords, e.g.

    lein eastwood "{:exclude-linters [:unlimited-use] :debug [:options :ns]}"

* `:all` - enable all debug messages.  This also enables showing the
  list of namespaces near the beginning of the output, before linting
  begins.
* `:options` - print the contents of the options map at several steps
  during startup.  May be useful to debug where options are coming
  from.
* `:config` - print the names of Eastwood config files just before
  they are read.
* `:time` - print messages about the elapsed time taken during
  analysis, and for each individual linter.
* `:forms` - print the forms as read, before they are analyzed
* `:forms-pprint` - like `:forms` except pretty-print the forms
* `:ast` - print ASTs as forms are analyzed and `eval`d.  These can be
  quite long.
* `:progress` - show a brief debug message after each top-level form
  is read
* `:compare-forms` - print all forms as read to a file
  `forms-read.txt`, all forms after being analyzed to a file
  `forms-analyzed.txt`, and all forms after being read, analyzed into
  an AST, and converted back into a form from the AST, to a file
  `forms-emitted.txt`.
* `:ns` - print the initial set of namespaces loaded into the Clojure
  run-time at the beginning of each file being linted.  (TBD: it used
  to do the following, but this needs to be reimplemented if it is
  desired: "and then after each top level form print any changes to
  that list of loaded namespaces (typically as the result of
  evaluating a `require` or `use` form).")
* `:var-info` - print some info about Vars that exist in various
  namespaces, and how many of them have data describing them in
  Eastwood's `var-info.edn` resource file, and how many do not.
  Useful when new releases of Clojure are made that add new Vars, to
  determine which ones Eastwood does not know about yet.


### Eastwood config files

Starting with version 0.2.1, Eastwood `eval`s several config files in
its internal resources.  You can see the latest versions
[here](https://github.com/jonase/eastwood/tree/master/resource/eastwood/config).
It also supports command line options to change which of these files
are read, or to read user-written config files.

Currently Eastwood supports config files that contain code to
selectively disable warnings of some linters.  For example, consider
this expression from config file `clojure.clj`:

```clojure
(disable-warning
 {:linter :suspicious-expression
  ;; specifically, those detected in function suspicious-macro-invocations
  :for-macro 'clojure.core/let
  :if-inside-macroexpansion-of #{'clojure.core/when-first}
  :within-depth 6
  :reason "when-first with an empty body is warned about, so warning about let with an empty body in its macroexpansion is redundant."})
```

Eastwood would normally report a `:suspicious-expression` warning if
it encounters a form `(let [x y])`, because the `let` has an empty
body.  It does so even if the `let` is the result of expanding some
other macro.

However, if such a `let` occurs because of a macro expansion of the
expression `(when-first [x y])`, the warning for the `let` will be
suppressed.  Eastwood already warns about the `(when-first [x y])`
because it has an empty body, so warning about the `let` would be
redundant.

The exact configurations supported are not documented yet, but will be
in the documentation for each linter.

You can specify the key `:builtin-config-files` in the options map to
override the built-in config files read.  It defaults to
`["clojure.clj" "clojure-contrib.clj" "third-party-libs.clj"]`.  All
such file names are only looked for in Eastwood's built-in config
files.

Similarly you can specify `:config-files` in the options map to give
additional files to read.  These are file names that can be anywhere
in your file system, specified as strings, or if Eastwood is invoked
from the REPL, anything that can be passed to
`clojure.java.io/reader`.


### Running Eastwood in a REPL

If you use Leiningen, merge this into your project's `project.clj`
file first:

```clojure
:profiles {:dev {:dependencies [[kirasystems/eastwood "0.2.5" :exclusions [org.clojure/clojure]]]}}
```

If you use a different build tool, you will need to add the dependency
above in the manner appropriate for it.  See
[Clojars](https://clojars.org/kirasystems/eastwood) for Gradle and Maven
syntax.

From within your REPL, there are two different functions you may call,
depending upon the kind of results you want.

* `eastwood` prints output similar to when you run Eastwood from the
  Leiningen command line, but it does not exit the JVM when it
  finishes.

* `lint` returns a map containing data structures describing any
  warnings or errors encountered.  This result is likely more useful
  for further processing by editors and IDEs.  For example, file
  names, line numbers, and column numbers are all available
  separately, requiring no parsing of strings containing those things
  combined together.  See the doc string of `eastwood.lint/lint` for
  details of the return value.

```clojure
(require '[eastwood.lint :as e])

;; Replace the values of :source-paths and :test-paths with whatever
;; is appropriate for your project.  You may omit them, and then the
;; default behavior is to search all directories in your Java
;; classpath, and their subdirectories recursively, for Clojure source
;; files.
(e/eastwood {:source-paths ["src"] :test-paths ["test"]})

(e/lint {:source-paths ["src"] :test-paths ["test"]})
```

All of the same options that can be given on a Leiningen command line
may be used in the map argument of the `eastwood` and `lint`
functions.

There is a `:callback` key that can be added to the argument map for
the `eastwood` function.  Its value is a callback function that gives
you significant control of where warning and error messages appear --
by default these all appear on the writer `*out*`.  This callback
function should not be overridden for the `lint` function, since
`lint` uses it in its implementation.

There is no documentation for this callback function yet.  You are
welcome to read Eastwood source code to see examples of how to write
one, but note that this is alpha-status code that will likely have API
changes in future Eastwood versions.


#### Warnings about using Eastwood in a REPL

Eastwood behaves similarly to `clojure.core/require` while performing
its analysis, in that it loads your code.  In particular, Eastwood
does these things:

* Reads and analyzes the source code you specify.
* Generates _new_ forms from the analysis results.  Note: if there are
  bugs, these new forms might not be identical to the original source
  code.
* Calls `eval` on the generated forms.

Hopefully you can see from this that Eastwood bugs, especially in the
portion up to generating new forms to be evaluated, could lead to
incorrect Clojure code being loaded into a running JVM.

It would be foolhardy to run Eastwood in a JVM running a live
production system.  We recommend that you use a different JVM process
for Eastwood than the one where you do your ongoing testing and
development work.

When reporting problems with Eastwood when run from the REPL, please
reproduce it in as few steps as possible after starting a new JVM
process, and include those steps in your problem report.

Running Eastwood from the REPL more than once in the same JVM process
requires you to manage your namespaces manually.  Eastwood will not
force the removal of any namespaces, and I would guess if there are
any issues from reloading a namespace that is already loaded with
protocols, `deftype`, etc. then they are yours to deal with.

Stuart Sierra's [component](https://github.com/stuartsierra/component)
library and workflow might be helpful in automatically removing old
versions of namespaces from a JVM process.  If you have instructions
that you have used with Eastwood and component or a similar tool,
please file a GitHub issue so they can be included here.


### Editor Support

See the section for the development environment you use for specific
instructions:

* [Eclipse+Counterclockwise](#eclipse--counterclockwise)
* [Emacs+Cider](#emacs--cider)
* [Emacs](#emacs)
* [Vim](#vim)

Below are instructions to create a text file of warnings that can be
useful for at least the Emacs and Vim cases, but are not needed for
the others.

As of Eastwood version 0.2.0, the new default warning message format
is no longer a Clojure map, but lines of the form:

    <file>:<line>:<col>: <linter> <msg>

You can still get the older map format by setting the option
`:warning-format` to `:map-v2` (`:map-v1` is the style used in
Eastwood 0.1.4 and earlier).  The new default format can be specified
explicitly by setting `:warning-format` to `:location-list-v1`.

You can put only the warning lines into a file using the option `:out`
followed by a file name in a double-quoted string, or when running
Eastwood from the REPL, anything convertible to a writer via
`clojure.java.io/writer`.

Note: If you try to mix reflection warnings from the Clojure compiler
in such a file, those messages contain relative path names from a
directory on your classpath, whereas Eastwood warnings contain
relative path names from the current directory.  For example, if the
directory `src` is on your classpath, as it is by default in Leiningen
projects, then Eastwood warnings will contain file names like
`src/your/ns/core.clj`, whereas reflection warnings will not have the
`src/` at the beginning.  If someone knows a good way to mingle
Eastwood and Clojure messages in the same file, please open a GitHub
issue for Eastwood.


### Eclipse + Counterclockwise

As of Dec 2014 this [Eastwood
plugin](https://github.com/laurentpetit/ccw-plugin-eastwood) for
[Eclipse+Counterclockwise](https://github.com/laurentpetit/ccw) was

changes.md
# Change log for Eastwood

## Changes from version 0.2.4 to 0.2.5

Internal enhancements:

* Update version of leinjacker for :managed-dependencies support.

## Changes from version 0.2.3 to 0.2.4

No new linters.  Added initial support for .cljc files and reader
conditionals.

* Read .cljc files in addition to .clj files when scanning namespaces.

* Handle reader conditionals by always parsing the :clj branch.

* Allow ClojureScript-specific libspec entries, such as `:include-macros
  true`.


Internal enhancements:

* Updated versions of tools.analyzer, tools.reader, tools.namespace,
  core.memoize, and clojure.java.classpath.

* Updated documentation, cleaned up comments, and removed unused
  require statements.


## Changes from version 0.2.2 to 0.2.3

No new linters.  The only difference with 0.2.2 is a few bug fixes:

* Ignore the contents of any file `data_readers.clj` in any directory
  of your classpath.  Earlier versions of Eastwood would lint their
  contents, and by default complain that there was no `ns` form at the
  beginning.  There should not be an `ns` form in such a file, and
  Eastwood no longer issues such incorrect warnings.  Issue
  [#172](https://github.com/jonase/eastwood/issues/172).

* Eliminate an exception caused by Eastwood's `unused-ret-vals` and
  `unused-ret-vals-in-try` linters that could occur if a Java method
  cannot be resolved.  Issue
  [#173](https://github.com/jonase/eastwood/issues/173).


## Changes from version 0.2.1 to 0.2.2

New linter:

* New linter `:wrong-pre-post` that warns about several kinds of wrong
  or suspicious preconditions or postconditions in a function.
  [link](https://github.com/jonase/eastwood#keyword-typos).  Issue
  [#89](https://github.com/jonase/eastwood/issues/89).


Other enhancements and bug fixes:

* Updates to support Clojure 1.8.0-RC1.

* Picture of Clint Eastwood in the README.

* Warn when a record field and function name are identical.  This was
  actually implemented in an earlier version of Eastwood, but I did
  not notice that until now.  Issue
  [#55](https://github.com/jonase/eastwood/issues/55).

* Several fixes to the bad-arglists linter.

* Change keyword-typos linter so that it no longer warns if the only
  difference between two keywords is the presence or absence of an
  initial underscore character.  Issue
  [#163](https://github.com/jonase/eastwood/issues/163).

* If you use the APIs to invoke Eastwood that return a map of info
  about each warning, there is a new :warning-details-url key that
  gives a link to the Eastwood documentation giving more detail about
  each type of warning.  Issue
  [#105](https://github.com/jonase/eastwood/issues/105).

* Eliminate exception thrown when linting byte-streams library.  Issue
  [#120](https://github.com/jonase/eastwood/issues/120).

* Disable unused-ret-vals warnings inside Potemkin's import-vars
  macro.  Issue [#135](https://github.com/jonase/eastwood/issues/135).


Internal enhancements:

* Correct the output put into debug file forms-emitted.txt.  Issue
  [#136](https://github.com/jonase/eastwood/issues/136).

* Updated versions of tools.analyzer, tools.analyzer.jvm and several
  other libraries used by Eastwood.  The current version of these
  should always be visible in the file
  [clone.sh](https://github.com/jonase/eastwood/blob/master/copy-deps-scripts/clone.sh)


There were changes in the way the 'defn' macro was implemented in
Clojure 1.8.0-RC1 that caused the tools.analyzer ASTs (abstract syntax
trees) to differ when analyzing Clojure code.  Eastwood 0.2.1 and
earlier did not gracefully handle those changes, causing the warnings
issued for the same Clojure project to differ significantly when using
Clojure 1.8.0-RC1 instead of an earlier version of Clojure, most
noticeably for misplaced doc strings and wrong tag warnings.
Hopefully all of these issues have been eliminated with Eastwood
0.2.2.


## Changes from version 0.2.0 to 0.2.1

If you use Emacs+Cider or Eclipse+Counterclockwise development
environments, there are now add-ons that integrate Eastwood warnings.
See https://github.com/jonase/eastwood#editor-support

New linters, and new good warnings from existing linters:

* New linter `:wrong-ns-form` that warns about several kinds of wrong
  or suspicious `:require` or `:use` subforms inside `ns` forms.
  Issue [#85](https://github.com/jonase/eastwood/issues/85),
  [#98](https://github.com/jonase/eastwood/issues/98)

* `:suspicious-expression` linter now warns about trivial uses of more
  `clojure.core` macros then before.

Fewer unwanted warnings, via logic enhancements or configuration
options:

* Several linters now have configuration options to disable their
  warnings based upon whether the warnings occur inside of a
  macroexpansion of a particular macro.  By default, Eastwood loads
  several config files worth of such disabling options for the linters
  `:constant-test`, `:redefd-vars`, `:suspicious-expression`, and
  `:unused-ret-vals` that prevent them from generating many unwanted
  warning messages, at least when certain macros are used, such as
  those in `core.contracts`, `core.match`, `core.typed`, Korma,
  Carmine, Timbre, Instaparse, and Schema.  Eastwood users may write
  their own config files to disable more warnings.
  Issue [#45](https://github.com/jonase/eastwood/issues/45),
  [#96](https://github.com/jonase/eastwood/issues/96),
  [#108](https://github.com/jonase/eastwood/issues/108),
  [#111](https://github.com/jonase/eastwood/issues/111),
  [#118](https://github.com/jonase/eastwood/issues/118),
  [#122](https://github.com/jonase/eastwood/issues/122),
  [#123](https://github.com/jonase/eastwood/issues/123)

* The `:wrong-arity` linter now generates nearly no unwanted warnings
  when you use the `java.jdbc` and Hiccup libraries.  Those libraries
  modify the `:arglists` key in metadata of some of their functions
  and macros for documentation purposes, but in a way that fooled

project.clj
(defproject kirasystems/eastwood "0.2.6-SNAPSHOT"
  :description "A Clojure lint tool"
  :url "https://github.com/kirasystems/eastwood"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :global-vars {*warn-on-reflection* true
                ;;*unchecked-math* :warn-on-boxed
               }
  :dependencies [
                 [org.clojure/clojure "1.5.1"]
                 ;[org.clojure/clojure "1.6.0"]
                 ;[org.clojure/clojure "1.7.0"]
                 ;[org.clojure/clojure "1.8.0"]
                 [org.clojars.brenton/google-diff-match-patch "0.1"]
                 [org.ow2.asm/asm-all "4.2"]]
  :profiles {:dev {:dependencies [[org.clojure/tools.macro "0.1.2"]
                                  [jafingerhut/dolly "0.1.0"]]
                   ;; I want the namespaces defined in files in
                   ;; "cases" and its subdirectories to be accessible
                   ;; during 'lein test' in the classpath, but they
                   ;; themselves should not be eval'd by 'lein test'.
                   ;; I also want those namespaces available for
                   ;; one-off tests at the command line like "lein
                   ;; eastwood '{:namespaces [testcases.testtest]}'",
                   ;; and for that, the :test profile does not work,
                   ;; so put them in :dev
                   :resource-paths [ "cases" ]}
             :test {:dependencies [
                                   ;; Needed for issue-173-test
                                   [commons-io "2.4"]
                                   ]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.1"]]}
             :1.6 {:dependencies [[org.clojure/clojure "1.6.0"]]}
             :1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
             :1.9 {:dependencies [[org.clojure/clojure "1.9.0-alpha16"]]}}
  ;; Note: comment out the following line if you want to do 'lein
  ;; test' and get a Clojure version later than 1.5.1, even if you
  ;; have an explicit org.clojure/clojure in the :dependencies
  ;; Also: Having the following line uncommented causes a problem when
  ;; doing 'lein test', I think because of a different version of
  ;; core.cache or core.memoize on which the latest tools.analyzer.jvm
  ;; 0.1.0-SNAPSHOT depends, versus what Leiningen itself probably
  ;; depends on.
;;  :eval-in-leiningen true
  ;; Eastwood may work with earlier Leiningen versions, but this is
  ;; close to the earliest version that it was most tested with.
  :min-lein-version "2.3.0"
  :resource-paths ["resource"]

  :repositories [["releases" {:url "https://clojars.org/repo"
                              :sign-releases false
                              :username :env
                              :password :env}]]
  ;;; Don't sign the tag, and don't deploy
  :release-tasks [["vcs" "assert-committed"]
                  ["change" "version" "leiningen.release/bump-version" "release"]
                  ["vcs" "commit"]
                  ["vcs" "tag" "--no-sign"]
                  ["change" "version" "leiningen.release/bump-version"]
                  ["vcs" "commit"]
                  ["vcs" "push"]])
